### ISSUE TITLE
Using MySqlGrammar generator SelectStatement unrecognizable Code

### ISSUE THREAD

**Issue by xlw712** [renatahodovan/grammarinator#25]
exec below:
`grammarinator-generate -l example/test/MySqlUnlexer.py -p example/test/MySqlUnparser.py -r  selectStatement -n 100  -d 20 -o tests/data/test_%d.txt -t grammarinator.runtime.simple_space_transformer`

return some data
`
( ( SELECT SQL_CALC_FOUND_ROWS  *  ORDER BY ! ! `𝍠`        DESC  , @JTOV2B2 := `🃑𐱉🥳`     NOT RLIKE @97     >  @"" := `ཉলא`     IS NOT \N        )  )  UNION ( SELECT *  ORDER BY @M6    NOT REGEXP `𑊏`      RLIKE @8 := `ஂ`     LIKE @'꜀' := @.$      IS UNKNOWN  DESC  , FALSE      OR  @`` := `𐠀𑴈ຘ𛲉`       | |  2     <  ALL ( ( SELECT *   )   )  IS FALSE     LIMIT @@``   , @@I     )   ORDER BY @'' := 0     NOT REGEXP @"" := NULL    NOT RLIKE @@45T     LIKE NULL    NOT REGEXP @@LWD       REGEXP @VF    NOT RLIKE @'' := \N     <  ANY ( ( ( ( SELECT STRAIGHT_JOIN  *   )  )  )  LOCK IN SHARE MODE   )  BETWEEN @`` := @@RT    < =  SOME ( ( SELECT SQL_CACHE  *   )  LOCK IN SHARE MODE   )  < =  @'''\Z' := @@K.    IS NOT NULL    AND NULL    SOUNDS LIKE @'' := NOT NULL     < =  NOT NULL    NOT BETWEEN @_    AND @"" := @W        IS FALSE  DESC  , BINARY CURRENT_TIME      < =  SOME ( ( ( SELECT SQL_CALC_FOUND_ROWS  HIGH_PRIORITY  SQL_BIG_RESULT  CURDATE  ( )    LIMIT 0   , 2    INTO `୞`   , @.    )  )  FOR UPDATE   )  IS NOT TRUE
`
Some special symbols appear,Is there any way to specify the resulting code?
please help me

**Comment by renatahodovan** [comment]
@xlw712 I think this is the same issue as #19. In short, there are two ways of specifying to generate only printable characters:

1) You should add an extra entry into the option section of your grammar which defines how dots (`.`) should be handled. `any_ascii_char` or `any_ascii_letter` will work. [example](https://github.com/renatahodovan/grammarinator/blob/master/examples/grammars/HTMLParser.g4#L36)

2) If you don't want to edit your grammar, then you can achive the same through cli: `grammarinator-process -Ddot=any_ascii_letter ...`

**Comment by xlw712** [comment]
> @xlw712 I think this is the same issue as #19. In short, there are two ways of specifying to generate only printable characters:
> 
> 1. You should add an extra entry into the option section of your grammar which defines how dots (`.`) should be handled. `any_ascii_char` or `any_ascii_letter` will work. [example](https://github.com/renatahodovan/grammarinator/blob/master/examples/grammars/HTMLParser.g4#L36)
> 2. If you don't want to edit your grammar, then you can achive the same through cli: `grammarinator-process -Ddot=any_ascii_letter ...`

 Thank you for your prompt reply,I go to try ~~

**Comment by xlw712** [comment]
Hi  I tried the second method 
`grammarinator-process -Ddot=any_ascii_letter /Users/xiaoliwen/code/mirror/mirror/src/main/g4/MySqlLexer.g4 /Users/xiaoliwen/code/mirror/mirror/src/main/g4/MySqlParser.g4 -o fuzzer `
but output error
`usage: grammarinator-process [-h] [--antlr FILE] [--no-actions]
                             [--encoding ENC] [--lib DIR] [--disable-cleanup]
                             [--pep8] [--log-level LEVEL] [-o DIR] [--version]
                             FILE [FILE ...]
grammarinator-process: error: unrecognized arguments: -Ddot=any_ascii_letter`
What am I doing wrong ？

**Comment by xlw712** [comment]
My version is up to date 
```
pip3 list  | grep grammarinator
grammarinator          19.3
```

**Comment by xlw712** [comment]
> My version is up to date
> 
> ```
> pip3 list  | grep grammarinator
> grammarinator          19.3
> ```
grammarinator-process not exists  -D command

**Comment by xlw712** [comment]
Is this way common to all grammars？

**Comment by xlw712** [comment]
> Hi I tried the second method
> `grammarinator-process -Ddot=any_ascii_letter /Users/xiaoliwen/code/mirror/mirror/src/main/g4/MySqlLexer.g4 /Users/xiaoliwen/code/mirror/mirror/src/main/g4/MySqlParser.g4 -o fuzzer `
> but output error
> `usage: grammarinator-process [-h] [--antlr FILE] [--no-actions] [--encoding ENC] [--lib DIR] [--disable-cleanup] [--pep8] [--log-level LEVEL] [-o DIR] [--version] FILE [FILE ...] grammarinator-process: error: unrecognized arguments: -Ddot=any_ascii_letter`
> What am I doing wrong ？

haha   Because pip3 installed the version is too low,I'm ready to run
step process
`
grammarinator-process -Ddot=any_unicode_char   MySqlLexer.g4 MySqlParser.g4 -o fuzzer
`
step generate
`
grammarinator-generate MySqlGenerator.MySqlGenerator --sys-path fuzzer  -d 10 -n 10  --serializer grammarinator.runtime.simple_space_serializer  -r selectStatement
`

 but output data is still wrong

`
( SELECT DISTINCTROW  SQL_BUFFER_RESULT  * , `𑌩𝄧𑇡`      LIMIT @@J   , 3178982     )  UNION ALL ( SELECT *  LIMIT @@XSC     )   UNION ( SELECT SQL_BUFFER_RESULT  SQL_CALC_FOUND_ROWS  *  LIMIT @@``   OFFSET @''     )   UNION ( SELECT STRAIGHT_JOIN  SQL_CACHE  ALL  SQL_BUFFER_RESULT  SQL_SMALL_RESULT  SQL_NO_CACHE  *   )   UNION ALL ( SELECT `🂯`   . *    INTO @'\𔙇'  , `𑙅𐼨`   , @````  , @''''  , `ꢀ`   , @"𑌴੦\𝔻\𑍵""𐨶""\🂠""჆"     )  LIMIT 1   , 1 
`

**Comment by akosthekiss** [comment]
@xlw712 So, I guess you've managed to install via `python3 setup.py install`. And you still see those special characters. Could you share the relevant portion of your grammar that is responsible for those parts? Or would it be possible to share a modified minimal grammar that exhibits the same behavior?

**Comment by xlw712** [comment]
https://github.com/antlr/grammars-v4/tree/master/sql/mysql/Positive-Technologies

Thanks,I learned that it was the regular expression in the syntax that caused the error

**Comment by akosthekiss** [comment]
Was that a problem with a negated set (aka. not set)?

**Comment by xlw712** [comment]
> Was that a problem with a negated set (aka. not set)?

yes,I 'm sure

**Comment by akosthekiss** [comment]
@xlw712 A major charset rework has landed, which changes how (and when) inverted/negated/not sets are calculated. Could you try and see whether it solves your issue?

**Comment by xlw712** [comment]
> @xlw712 A major charset rework has landed, which changes how (and when) inverted/negated/not sets are calculated. Could you try and see whether it solves your issue?

Is the latest master branch?

**Comment by akosthekiss** [comment]
> > @xlw712 A major charset rework has landed, which changes how (and when) inverted/negated/not sets are calculated. Could you try and see whether it solves your issue?
> 
> Is the latest master branch?

Yes, latest changes are available from the master branch.

**Comment by xlw712** [comment]
> > > @xlw712 A major charset rework has landed, which changes how (and when) inverted/negated/not sets are calculated. Could you try and see whether it solves your issue?
> > 
> > 
> > Is the latest master branch?
> 
> Yes, latest changes are available from the master branch.

Ok, I go to try , late reply

**Comment by xlw712** [comment]
@akosthekiss  
haha , loooks good
`
pip3` list | grep grammarinator
grammarinator          19.3+53.gbfde275
`
There is no unrecognizable code in the data

**Comment by renatahodovan** [comment]
@xlw712 Thanks for the feedback! Fixed by 480f82ed2395e0155ba3ba24b3266843dd2a5fbb.
